### PR TITLE
add storage and authentication section

### PR DESCRIPTION
**Pull request by Cloudwu1**
## Description

Add storage max size, storage file size and authentication section to macOS swift and objectiveC apps.

before:

image

After:
image

### CODE REVIEW

File: SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift [hunk 1]
```diff
@@ -53,4 +93,38 @@ class AppCenterViewController : NSViewController {
...
+    @IBAction func singOutClicked(_ sender: NSButton) {
+        showSignInController(action: AuthenticationViewController.AuthAction.signout)
+    }
+
+    override func prepare(for segue: NSStoryboardSegue, sender: Any?) {
+        if let signInController = segue.destinationController as? AuthenticationViewController {
+            signInController.action = currentAction
+        }
+    }
+
```
Comment (MatkovIvan) [inline SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift]: Please format this file

File: SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        return url.queryItems?.first(where: { $0.name == queryParamaterName })?.value
+    }
+}
+
+class AuthenticationViewController: NSViewController, WKNavigationDelegate, MSAnalyticsAuthenticationProviderDelegate {
+
+    var onAuthDataReceived: ((_ token: String, _ userId: String, _ expiresAt: Date) -> Void)?
+
+    enum AuthAction {
+        case login, signout
```
Comment (MatkovIvan) [inline SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift]: login/logout
or
signin/signout

File: SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    override func loadView() {
+        super.loadView()
+        let configuration = WKWebViewConfiguration()
+        configuration.preferences.javaScriptEnabled = true
+        if #available(OSX 10.11, *) {
+            configuration.websiteDataStore = WKWebsiteDataStore.nonPersistent()
+        } else {
+            // Fallback on earlier versions
```
Comment (MatkovIvan) [inline SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift]: TODO? Will it be in this PR?

File: SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+                NSLog("Error while signing out: %@", error)
+            } else {
+                NSLog("Successfully signed out")
+            }
+            close()
+        }
+    }
+
+    func close() {
+        //self.dismiss(animated: true, completion: nil)
```
Comment (MatkovIvan) [inline SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift]: Remove if not needed

File: SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+                        NSLog("Refresh token error: \"\(error)\": \(errorDescription)")
+                        // Call the completion handler in the error case to send anonymous logs.
+                        completionHandler(nil, nil)
+                        return
+                    }
+                    let token = json["access_token"]! as! String
+                    let expiresIn = json["expires_in"]! as! Int64
+                    let userId = json["user_id"]! as! String
+                    NSLog("Successfully refreshed token for user: %@", userId)
+                    // Call the completion handler and pass in the updated token and expiryDate.
```
Comment (MatkovIvan) [inline SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift]: Empty line before comments

File: SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+                    completionHandler(token, Date().addingTimeInterval(Double(expiresIn)))
+                } catch let error as JSONError {
+                    NSLog("Error while preforming refresh request: %@", error.rawValue)
+                } catch let error as NSError {
+                    NSLog("Error while preforming refresh request: %@", error.localizedDescription)
+                }
+                }.resume()
```
Comment (MatkovIvan) [inline SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift]:
```suggestion
            }.resume()
```

File: SasquatchMac/SasquatchMacSwift/AppDelegate.swift [hunk 7]
```diff
@@ -59,6 +59,22 @@ class AppDelegate: NSObject, NSApplicationDelegate, MSCrashesDelegate, MSPushDel
...
+            DispatchQueue.main.async {
+                if success {
+                    let realSize = Int64(ceil(Double(storageMaxSize!) / Double(kMSStoragePageSize))) * Int64(kMSStoragePageSize)
+                    UserDefaults.standard.set(realSize, forKey: kMSStorageMaxSizeKey)
+                } else {
+                    // Remove invalid value.
+                    UserDefaults.standard.removeObject(forKey: kMSStorageMaxSizeKey)
+                }
+            }
+        })
```
Comment (MatkovIvan) [inline SasquatchMac/SasquatchMacSwift/AppDelegate.swift]: It looks like we use tab size as 2 spaces here. Please recheck formatting in your changes

File: SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift [hunk 8]
```diff
@@ -1,8 +1,12 @@
 import Cocoa
 
-class AppCenterViewController : NSViewController {
+// 100 MiB.
+let kMSDefaultDatabaseSize = 100 * 1024 * 1024
```
Comment (jwallra) [inline SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift]: Let's use the same default as other SDKs - 10MiB

File: SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift [hunk 9]
```diff
@@ -53,4 +93,38 @@ class AppCenterViewController : NSViewController {
     UserDefaults.standard.set(indexNumber, forKey: kMSStartTargetKey)
   }
 
+  //Storage Max Size
```
Comment (jwallra) [inline SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift]: nit: space after `//`

File: SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+            if let newUrl = URL(string: self.baseUrl + self.redirectEndpoint + "?" + url.fragment!) {
+                if let error = newUrl.valueOf("error") {
+                    NSLog("Error while signing in: %@", error)
+                    self.close()
+                } else {
+                    let refreshToken = newUrl.valueOf(self.refreshTokenParam)!
+                    if(!refreshToken.isEmpty) {
+                        self.refreshToken = refreshToken
+                        NSLog("Successfully signed in with user_id: %@", newUrl.valueOf("user_id")!)
+                        // Create a MSAnalyticsAuthenticationProvider and register as an MSAnalyticsAuthenticationProvider.
```
Comment (jwallra) [inline SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift]: new line before comment

File: SasquatchMac/SasquatchMacObjC/AppDelegate.m [hunk 11]
```diff
@@ -31,6 +31,24 @@ - (void)applicationDidFinishLaunching:(NSNotification *)notification {
...
+  if (storageMaxSize) {
+      [MSAppCenter setMaxStorageSize:storageMaxSize.integerValue
+                   completionHandler:^(BOOL success) {
+                       dispatch_async(dispatch_get_main_queue(), ^{
+                           if (success) {
+                               long realStorageSize = (long)(ceil([storageMaxSize doubleValue] / kMSStoragePageSize) * kMSStoragePageSize);
+                               [[NSUserDefaults standardUserDefaults] setObject:[NSNumber numberWithLong:realStorageSize]
+                                                                  forKey:kMSStorageMaxSizeKey];
+                           } else {
+                               // Remove invalid value.
```
Comment (jwallra) [inline SasquatchMac/SasquatchMacObjC/AppDelegate.m]: new line before comment

File: SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift [hunk 12]
```diff
@@ -1,8 +1,12 @@
 import Cocoa
 
-class AppCenterViewController : NSViewController {
+// 100 MiB.
```
Comment (ElektrojungeMS) [inline SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift]: As we're not on a mobile platform 100MiB seems to be an "okay" number but given that most macOS applications are a few 100s of MIB in size, setting our DB size to 100MiB seems a bit excessive. Why not keep it the same size as the other SDKs (which are 10MiB)?

File: SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+            let bodyString = redirectParam + clientIdParam + refreshParam + refreshToken + scopeParam
+            let data: Data = bodyString.data(using: String.Encoding.utf8)!
+            NSLog("Started refresh process")
+            session.uploadTask(with: request as URLRequest, from: data) { (data, response, error) in
+                defer {
+                    self.close()
+                }
+                do {
+                    guard let data = data else {
+                        // Call the completion handler in the error case to send anonymous logs.
```
Comment (ElektrojungeMS) [inline SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift]: Add a new line before the comment

File: SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift (hunk 1 above)
Comment (Cloudwu1) [reply SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift]: Fix in new commit.

File: SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift (hunk 2 above)
Comment (Cloudwu1) [reply SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift]: Fix in new commit.

File: SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift (hunk 3 above)
Comment (Cloudwu1) [reply SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift]: Fix in new commit.

File: SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift (hunk 4 above)
Comment (Cloudwu1) [reply SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift]: Fix in new commit.

File: SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift (hunk 5 above)
Comment (Cloudwu1) [reply SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift]: Fix in new commit.

File: SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift (hunk 6 above)
Comment (Cloudwu1) [reply SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift]: Fix in new commit.

File: SasquatchMac/SasquatchMacSwift/AppDelegate.swift (hunk 7 above)
Comment (Cloudwu1) [reply SasquatchMac/SasquatchMacSwift/AppDelegate.swift]: Fix in new commit.

File: SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift (hunk 8 above)
Comment (Cloudwu1) [reply SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift]: Fix in new commit.

File: SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift (hunk 10 above)
Comment (Cloudwu1) [reply SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift]: Fix in new commit.

File: SasquatchMac/SasquatchMacObjC/AppDelegate.m (hunk 11 above)
Comment (Cloudwu1) [reply SasquatchMac/SasquatchMacObjC/AppDelegate.m]: Fix in new commit.

File: SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift (hunk 12 above)
Comment (Cloudwu1) [reply SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift]: Fix in new commit.

File: SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift (hunk 13 above)
Comment (Cloudwu1) [reply SasquatchMac/SasquatchMac/ViewControllers/AuthenticationViewController.swift]: Fix in new commit.

File: SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift (hunk 9 above)
Comment (Cloudwu1) [reply SasquatchMac/SasquatchMac/ViewControllers/AppCenterViewController.swift]: Fix in new commit.

**Pull request merged**